src/flp/types/dp.rs

```diff
@@ -0,0 +1,327 @@
+use crate::dp::{distributions::PureDpDiscreteLaplace, DifferentialPrivacyStrategy};
+use crate::dp::{DifferentialPrivacyDistribution, DpError};
+use crate::field::{FftFriendlyFieldElement, Field128, Field64};
+use crate::flp::gadgets::{Mul, ParallelSumGadget};
+use crate::flp::types::{Histogram, SumVec};
+use crate::flp::{FlpError, TypeWithNoise};
+use crate::vdaf::xof::SeedStreamTurboShake128;
+use num_bigint::{BigInt, BigUint, TryFromBigIntError};
+use num_integer::Integer;
+use num_traits::{One, Pow};
+use rand::{distributions::Distribution, Rng, SeedableRng};
+
+// TODO(#1071): This is implemented for the concrete fields `Field64` and `Field128` in order to
+// avoid imposing the `BigInt: From<F::Integer>` bound on all callers. In the future, we may want to
+// impose this bound on `FieldElementWithInteger`, which would allow simple blanket implementations
+// here.
+impl<S> TypeWithNoise<PureDpDiscreteLaplace> for SumVec<Field64, S>
+where
+    S: ParallelSumGadget<Field64, Mul<Field64>> + Eq + 'static,
+{
+    fn add_noise_to_result(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [Self::Field],
+        _num_measurements: usize,
+    ) -> Result<(), FlpError> {
+        self.add_noise(
+            dp_strategy,
+            agg_result,
+            &mut SeedStreamTurboShake128::from_entropy(),
+        )
+    }
+}
+
+impl<S> TypeWithNoise<PureDpDiscreteLaplace> for SumVec<Field128, S>
+where
+    S: ParallelSumGadget<Field128, Mul<Field128>> + Eq + 'static,
+{
+    fn add_noise_to_result(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [Self::Field],
+        _num_measurements: usize,
+    ) -> Result<(), FlpError> {
+        self.add_noise(
+            dp_strategy,
+            agg_result,
+            &mut SeedStreamTurboShake128::from_entropy(),
+        )
+    }
+}
+
+impl<F, S> SumVec<F, S>
+where
+    F: FftFriendlyFieldElement,
+    BigInt: From<F::Integer>,
+    F::Integer: TryFrom<BigInt, Error = TryFromBigIntError<BigInt>>,
+{
+    fn add_noise<R>(
+        &self,
+        dp_strategy: &PureDpDiscreteLaplace,
+        agg_result: &mut [F],
+        rng: &mut R,
+    ) -> Result<(), FlpError>
+    where
+        R: Rng,
+    {
+        // Compute the l1-sensitivity of the aggregation function (assuming the substitution-DP
+        // model). The worst case is when one individual's measurement changes such that each vector
+        // element flips from 0 to 2^bits - 1, or vice versa. Then, the l1 distance from the initial
+        // query result to the new query result will be (2^bits - 1) * length.
+        let two = BigUint::from(2u64);
+        let bits = BigUint::from(self.bits);
+        let length = BigUint::from(self.len);
+        let sensitivity = (Pow::pow(two, &bits) - BigUint::one()) * length;
```

Comment:
Isn't this going to generate too much noise?  The sensitivity for each dimension is $2^b-1$, which requires IID noise for each dimension proportional to $2^b-1$, not $\ell\cdot2^b-1$.  Unless I'm misreading your IID function, that is.

Reply:
Separately, do you expect to support self.bits > 127?  Such a vector would be very expensive, but it would still be possible to do `let sensitivity = BigUint::from((1u128 << self.bits) - 1)`  or (clever) `let sensitivity = BigUint::from(u128::max >> 128.checked_sub(self.bits).ok_or(Error...)?)`.

Reply:
Yes, this would add noise computed based on a sensitivity of $\ell \cdot (2^b-1)$ to each dimension. I think this is correct as far as I can see, and I think this is the main reason why other VDAFs, when composed with DP, benefit from tighter constraints on measurements. (i.e. Prio3Histogram here, Prio3FixedPointBoundedL2VecSum previously, or PINE)

If we split up one $\epsilon$-DP Prio3SumVec query into $\ell$ many Prio3Sum queries, we would have a pure DP budget of $\epsilon/\ell$ for each. The sensitivity of Prio3Sum is $2^b-1$. Thus, we'd add noise drawn from $Lap_\mathbb{Z}((2^b-1)/(\epsilon/\ell))=Lap_\mathbb{Z}(\ell\cdot(2^b-1)/\epsilon)$, which is the same as we'd get from this code as-is.

Another way I thought about this: if the adversary happens to know a priori that each dimension of some client's measurement is perfectly correlated or anti-correlated to the rest, then we'd need that factor of $\ell$ to avoid effectively giving them free queries beyond our budget.

Reply:
We don't need to support self.bits > 127, because the output share would be able to wrap around Field128, and we haven't defined any bigger FFT-friendly fields as of yet, so that works.

Reply:
Just acknowledging that factor of $\ell$.  I've spent so long thinking about histograms, I'd forgotten how to think about vectors.  That's going to be a lot of noise, but ¯\\\_(ツ)\_/¯